File: swift-bridge/Sources/CoreImageBridge/BarcodeDescriptor.swift
import CoreImage
import Foundation

private func ci_qr_level(_ code: Int32) -> CIQRCodeDescriptor.ErrorCorrectionLevel? {
    switch code {
    case 0:
        return .levelL
    case 1:
        return .levelM
    case 2:
        return .levelQ
    case 3:
        return .levelH
    default:
        return nil
    }
}

private func ci_data_matrix_ecc(_ code: Int32) -> CIDataMatrixCodeDescriptor.ECCVersion? {
    switch code {
    case 0:
        return .v000
    case 50:
        return .v050
    case 80:
        return .v080
    case 100:
        return .v100
    case 140:
        return .v140
    case 200:
        return .v200
    default:
        return nil
    }
}

private func ci_descriptor_payload(_ descriptor: CIBarcodeDescriptor) -> Data {
    switch descriptor {
    case let qr as CIQRCodeDescriptor:
        return qr.errorCorrectedPayload
    case let aztec as CIAztecCodeDescriptor:
        return aztec.errorCorrectedPayload
    case let pdf as CIPDF417CodeDescriptor:
        return pdf.errorCorrectedPayload
    case let dataMatrix as CIDataMatrixCodeDescriptor:
        return dataMatrix.errorCorrectedPayload
    default:
        return Data()
    }
}

@_cdecl("ci_barcode_descriptor_new_qr")
public func ci_barcode_descriptor_new_qr(
    _ bytes: UnsafePointer<UInt8>?,
    _ len: Int,
    _ symbolVersion: Int,
    _ maskPattern: UInt8,
    _ errorCorrectionLevel: Int32,
    _ outDescriptor: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    ci_run(outError) {
        guard let bytes, len >= 0, let level = ci_qr_level(errorCorrectionLevel), let outDescriptor else {
            throw CIBridgeError.invalidArgument("invalid QR descriptor arguments")
        }
        let data = Data(bytes: bytes, count: len)
        guard let descriptor = CIQRCodeDescriptor(
            payload: data,
            symbolVersion: symbolVersion,
            maskPattern: maskPattern,
            errorCorrectionLevel: level
        ) else {
            throw CIBridgeError.nullResult("CIQRCodeDescriptor initializer returned nil")
        }
        outDescriptor.pointee = ci_retain(descriptor)
    }
}

@_cdecl("ci_barcode_descriptor_new_aztec")
public func ci_barcode_descriptor_new_aztec(
    _ bytes: UnsafePointer<UInt8>?,
    _ len: Int,
    _ isCompact: Bool,
    _ layerCount: Int,
    _ dataCodewordCount: Int,
    _ outDescriptor: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    ci_run(outError) {
        guard let bytes, len >= 0, let outDescriptor else {
            throw CIBridgeError.invalidArgument("invalid Aztec descriptor arguments")
        }
        let data = Data(bytes: bytes, count: len)
        guard let descriptor = CIAztecCodeDescriptor(
            payload: data,
            isCompact: isCompact,
            layerCount: layerCount,
            dataCodewordCount: dataCodewordCount
        ) else {
            throw CIBridgeError.nullResult("CIAztecCodeDescriptor initializer returned nil")
        }
        outDescriptor.pointee = ci_retain(descriptor)
    }
}

@_cdecl("ci_barcode_descriptor_new_pdf417")
public func ci_barcode_descriptor_new_pdf417(
    _ bytes: UnsafePointer<UInt8>?,
    _ len: Int,
    _ isCompact: Bool,
    _ rowCount: Int,
    _ columnCount: Int,
    _ outDescriptor: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    ci_run(outError) {
        guard let bytes, len >= 0, let outDescriptor else {
            throw CIBridgeError.invalidArgument("invalid PDF417 descriptor arguments")
        }
        let data = Data(bytes: bytes, count: len)
        guard let descriptor = CIPDF417CodeDescriptor(
            payload: data,
            isCompact: isCompact,
            rowCount: rowCount,
            columnCount: columnCount
        ) else {
            throw CIBridgeError.nullResult("CIPDF417CodeDescriptor initializer returned nil")
        }
        outDescriptor.pointee = ci_retain(descriptor)
    }
}

@_cdecl("ci_barcode_descriptor_new_data_matrix")
public func ci_barcode_descriptor_new_data_matrix(
    _ bytes: UnsafePointer<UInt8>?,
    _ len: Int,
    _ rowCount: Int,
    _ columnCount: Int,
    _ eccVersion: Int32,
    _ outDescriptor: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    ci_run(outError) {
        guard let bytes, len >= 0, let ecc = ci_data_matrix_ecc(eccVersion), let outDescriptor else {
            throw CIBridgeError.invalidArgument("invalid Data Matrix descriptor arguments")
        }
        let data = Data(bytes: bytes, count: len)
        guard let descriptor = CIDataMatrixCodeDescriptor(
            payload: data,
            rowCount: rowCount,
            columnCount: columnCount,
            eccVersion: ecc
        ) else {
            throw CIBridgeError.nullResult("CIDataMatrixCodeDescriptor initializer returned nil")
        }
        outDescriptor.pointee = ci_retain(descriptor)
    }
}

@_cdecl("ci_barcode_descriptor_kind")
public func ci_barcode_descriptor_kind(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let descriptor: CIBarcodeDescriptor = ci_borrow(handle) else { return -1 }
    switch descriptor {
    case is CIQRCodeDescriptor:
        return 0
    case is CIAztecCodeDescriptor:
        return 1
    case is CIPDF417CodeDescriptor:
        return 2
    case is CIDataMatrixCodeDescriptor:
        return 3
    default:
        return -1
    }
}

@_cdecl("ci_barcode_descriptor_payload_base64")
public func ci_barcode_descriptor_payload_base64(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let descriptor: CIBarcodeDescriptor = ci_borrow(handle) else { return nil }
    return ci_string(ci_descriptor_payload(descriptor).base64EncodedString())
}

@_cdecl("ci_barcode_descriptor_details_json")
public func ci_barcode_descriptor_details_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let descriptor: CIBarcodeDescriptor = ci_borrow(handle) else { return ci_string("{}") }
    var details: [String: Any] = [
        "kind": ci_barcode_descriptor_kind(handle),
        "payloadBase64": ci_descriptor_payload(descriptor).base64EncodedString(),
    ]
    if let qr = descriptor as? CIQRCodeDescriptor {
        details["symbolVersion"] = qr.symbolVersion
        details["maskPattern"] = qr.maskPattern
        details["errorCorrectionLevel"] = qr.errorCorrectionLevel.rawValue
    }
    if let aztec = descriptor as? CIAztecCodeDescriptor {
        details["isCompact"] = aztec.isCompact
        details["layerCount"] = aztec.layerCount
        details["dataCodewordCount"] = aztec.dataCodewordCount
    }
    if let pdf = descriptor as? CIPDF417CodeDescriptor {
        details["isCompact"] = pdf.isCompact
        details["rowCount"] = pdf.rowCount
        details["columnCount"] = pdf.columnCount
    }
    if let dataMatrix = descriptor as? CIDataMatrixCodeDescriptor {
        details["rowCount"] = dataMatrix.rowCount
        details["columnCount"] = dataMatrix.columnCount
        details["eccVersion"] = dataMatrix.eccVersion.rawValue
    }
    return ci_string(ci_json_string(from: details) ?? "{}")
}

File: src/barcode_descriptor.rs
use core::ffi::c_void;
use core::fmt;
use core::ptr;

use crate::ffi;
use crate::util::{status_result, take_owned_string};
use crate::CIError;

/// Barcode descriptor kinds supported by Core Image.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CIBarcodeDescriptorKind {
/// Mirrors the `CoreImage` framework case `QrCode`.
    QrCode,
/// Mirrors the `CoreImage` framework case `Aztec`.
    Aztec,
/// Mirrors the `CoreImage` framework case `Pdf417`.
    Pdf417,
/// Mirrors the `CoreImage` framework case `DataMatrix`.
    DataMatrix,
/// Mirrors the `CoreImage` framework case `Unknown`.
    Unknown,
}

/// QR code error-correction levels.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CIQRCodeErrorCorrectionLevel {
/// Mirrors the `CoreImage` framework case `L`.
    L,
/// Mirrors the `CoreImage` framework case `M`.
    M,
/// Mirrors the `CoreImage` framework case `Q`.
    Q,
/// Mirrors the `CoreImage` framework case `H`.
    H,
}

impl CIQRCodeErrorCorrectionLevel {
    const fn code(self) -> i32 {
        match self {
            Self::L => 0,
            Self::M => 1,
            Self::Q => 2,
            Self::H => 3,
        }
    }
}

/// Data Matrix ECC versions.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CIDataMatrixCodeECCVersion {
/// Mirrors the `CoreImage` framework case `V000`.
    V000,
/// Mirrors the `CoreImage` framework case `V050`.
    V050,
/// Mirrors the `CoreImage` framework case `V080`.
    V080,
/// Mirrors the `CoreImage` framework case `V100`.
    V100,
/// Mirrors the `CoreImage` framework case `V140`.
    V140,
/// Mirrors the `CoreImage` framework case `V200`.
    V200,
}

impl CIDataMatrixCodeECCVersion {
    const fn code(self) -> i32 {
        match self {
            Self::V000 => 0,
            Self::V050 => 50,
            Self::V080 => 80,
            Self::V100 => 100,
            Self::V140 => 140,
            Self::V200 => 200,
        }
    }
}

/// A machine-readable barcode descriptor.
pub struct CIBarcodeDescriptor {
    ptr: *mut c_void,
}

impl Drop for CIBarcodeDescriptor {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::ci_object_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

impl Clone for CIBarcodeDescriptor {
    fn clone(&self) -> Self {
        Self {
            ptr: unsafe { ffi::ci_object_retain(self.ptr) },
        }
    }
}

impl fmt::Debug for CIBarcodeDescriptor {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("CIBarcodeDescriptor")
            .field("ptr", &self.ptr)
            .field("kind", &self.kind())
            .finish_non_exhaustive()
    }
}

impl CIBarcodeDescriptor {
    pub(crate) const unsafe fn from_raw(ptr: *mut c_void) -> Self {
        Self { ptr }
    }

    fn from_non_null(ptr: *mut c_void, kind: &str) -> Self {
        assert!(!ptr.is_null(), "{kind} returned nil");
        unsafe { Self::from_raw(ptr) }
    }

/// Mirrors the `CoreImage` framework constant `fn`.
    pub const fn as_ptr(&self) -> *mut c_void {
        self.ptr
    }

/// Calls the `CoreImage` framework counterpart for `qr_code`.
    pub fn qr_code(
        payload: &[u8],
        symbol_version: isize,
        mask_pattern: u8,
        error_correction_level: CIQRCodeErrorCorrectionLevel,
    ) -> Result<Self, CIError> {
        let mut descriptor = ptr::null_mut();
        let mut error = ptr::null_mut();
        let status = unsafe {
            ffi::ci_barcode_descriptor_new_qr(
                payload.as_ptr(),
                payload.len(),
                symbol_version,
                mask_pattern,
                error_correction_level.code(),
                &mut descriptor,
                &mut error,
            )
        };
        unsafe { status_result(status, error)? };
        Ok(Self::from_non_null(descriptor, "CIQRCodeDescriptor"))
    }

/// Calls the `CoreImage` framework counterpart for `aztec_code`.
    pub fn aztec_code(
        payload: &[u8],
        is_compact: bool,
        layer_count: isize,
        data_codeword_count: isize,
    ) -> Result<Self, CIError> {
        let mut descriptor = ptr::null_mut();
        let mut error = ptr::null_mut();
        let status = unsafe {
            ffi::ci_barcode_descriptor_new_aztec(
                payload.as_ptr(),
                payload.len(),
                is_compact,
                layer_count,
                data_codeword_count,
                &mut descriptor,
                &mut error,
            )
        };
        unsafe { status_result(status, error)? };
        Ok(Self::from_non_null(descriptor, "CIAztecCodeDescriptor"))
    }

/// Calls the `CoreImage` framework counterpart for `pdf417_code`.
    pub fn pdf417_code(
        payload: &[u8],
        is_compact: bool,
        row_count: isize,
        column_count: isize,
    ) -> Result<Self, CIError> {
        let mut descriptor = ptr::null_mut();
        let mut error = ptr::null_mut();
        let status = unsafe {
            ffi::ci_barcode_descriptor_new_pdf417(
                payload.as_ptr(),
                payload.len(),
                is_compact,
                row_count,
                column_count,
                &mut descriptor,
                &mut error,
            )
        };
        unsafe { status_result(status, error)? };
        Ok(Self::from_non_null(descriptor, "CIPDF417CodeDescriptor"))
    }

/// Calls the `CoreImage` framework counterpart for `data_matrix_code`.
    pub fn data_matrix_code(
        payload: &[u8],
        row_count: isize,
        column_count: isize,
        ecc_version: CIDataMatrixCodeECCVersion,
    ) -> Result<Self, CIError> {
        let mut descriptor = ptr::null_mut();
        let mut error = ptr::null_mut();
        let status = unsafe {
            ffi::ci_barcode_descriptor_new_data_matrix(
                payload.as_ptr(),
                payload.len(),
                row_count,
                column_count,
                ecc_version.code(),
                &mut descriptor,
                &mut error,
            )
        };
        unsafe { status_result(status, error)? };
        Ok(Self::from_non_null(
            descriptor,
            "CIDataMatrixCodeDescriptor",
        ))
    }

/// Calls the `CoreImage` framework counterpart for `kind`.
    pub fn kind(&self) -> CIBarcodeDescriptorKind {
        match unsafe { ffi::ci_barcode_descriptor_kind(self.ptr) } {
            0 => CIBarcodeDescriptorKind::QrCode,
            1 => CIBarcodeDescriptorKind::Aztec,
            2 => CIBarcodeDescriptorKind::Pdf417,
            3 => CIBarcodeDescriptorKind::DataMatrix,
            _ => CIBarcodeDescriptorKind::Unknown,
        }
    }

/// Calls the `CoreImage` framework counterpart for `payload_base64`.
    pub fn payload_base64(&self) -> String {
        unsafe { take_owned_string(ffi::ci_barcode_descriptor_payload_base64(self.ptr)) }
            .unwrap_or_default()
    }

/// Calls the `CoreImage` framework counterpart for `details_json`.
    pub fn details_json(&self) -> String {
        unsafe { take_owned_string(ffi::ci_barcode_descriptor_details_json(self.ptr)) }
            .unwrap_or_else(|| "{}".to_string())
    }
}
